kitsune/search/v2/base.py

```diff
@@ -22,18 +22,19 @@ def prepare(cls, instance):
 
             # This will allow child classes to have their own methods in the form of prepare_field
             prepare_method = getattr(obj, "prepare_{}".format(f), None)
-            # Ovewrite the value if there is a specific method implented for the specific field
-            if prepare_method is not None:
-                value = prepare_method(instance)
-            else:
-                value = getattr(instance, f)
-
-            # Assign values to each field. Assign a dictionary to Objects
-            if isinstance(doc_mapping.resolve_field(f), field.Object):
+            value = obj.get_field_value(f, instance, prepare_method)
+
+            # Assign values to each field.
+            field_type = doc_mapping.resolve_field(f)
+            if isinstance(field_type, field.Object) and not (
+                isinstance(value, InnerDoc)
+                or (isinstance(value, list) and isinstance((value or [None])[0], InnerDoc))
```

Comment:
Could you please elaborate a bit on the different scenarios that you are checking against here?

Reply:
assuming we have 5 values, all being inserted into an Object field:

1. `"hello hello"` - this is a string (or, not an `InnerDoc`), so we need to apply the locale logic, the conditional matches here
2. `UserInnerDoc(username="foobar")` - this is an InnerDoc, we don't want to apply any locale logic, the conditional doesn't match
3. `[UserInnerDoc(username="foobar")]` - this is an InnerDoc in a list, we don't want to apply any locale logic, however `value` is not an `InnerDoc` (being a `list`), so we need the logic after the `or` to check the type of the first instance if we have a list, the conditional doesn't match
4. `["hello hello"]` - we do want to apply locale logic here, the conditional matches
5. `[]` - we don't really care what happens here, as long as nothing throws an exception, which is why we need the `(value or [None])[0]`. (I think) the conditional will match this.

Reply:
I still believe that having something like the snippet (re-adding it here) that I added before should be the way to go for now, especially without any other `InnerDoc`s.

eg
```python
 if isinstance(field_type, field.Object) and "en-US" in field_type._mapping:
    if locale := obj.prepare_locale(instance) and locale in field_type._mapping:
       if value:
           obj[f] = {locale:value}
    else:
       other_type_of_objects
```
For example the `isinstance(value, InnerDoc)` is redundant here. When a field is of type `Object` without and `doc_class` as we have in the `SumoLocaleField` then it's initialized as an InnerDoc. So basically here we are saying if something is of type Object(InnerDoc) and not InnerDoc. The above snippet is not optimum but I would defer dealing with this issue - if it ever becomes one - when we have the need for it.

Reply:
Also this way we are treating all `String` in the object as a locale field which should not be the case. Ideally we should get somehow the type of `SumoLocale` field. If this is not possible, probably the second best would be to look at the data. Thoughts?

Reply:
> So basically here we are saying if something is of type Object(InnerDoc) and not InnerDoc.

Yes, that's the idea - if we have a mismatch in types between our field and value then we know that we're dealing with a SumoLocale field.

> Also this way we are treating all String in the object as a locale field which should not be the case.

I'm not sure there are any cases other than the SumoLocale field where a prepare_some_object function would return a String (or any other non-InnerDoc).

Regardless this isn't needed here yet, I'll add a non-list-handling version in the community PR. Anyway we may want to change this logic again later once we implement the parent->child indexing of kb articles.